src/dhtproto/client/helper/ChannelSerializer.d

```diff
@@ -0,0 +1,354 @@
+/*******************************************************************************
+
+    DHT channel serializer helper.
+
+    Helper class template for dumping a cached or mirrored channel from a map
+    (of some kind) in memory to disk, then loading it again at startup. The
+    dumping can be done in either a blocking way (e.g. at application shutdown)
+    or asynchronously via `fork` (e.g. periodically while the application is
+    running).
+
+    Notes:
+        * Blocking dump: Dumps the map without forking. If an asynchronous dump
+          is already in progress, the call will block until the dump is
+          finished, then start it again.
+        * Asynchronous dump: When an asynchronous dump is in progress, further
+          calls to dump will not start a second, clashing dump; the dump method
+          will simply return.
+
+    Copyright:
+        Copyright (c) 2018 dunnhumby Germany GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module dhtproto.client.helper.ChannelSerializer;
+
+import ocean.transition;
+import ocean.io.device.File;
+import ocean.io.serialize.SimpleStreamSerializer;
+import ocean.io.stream.Buffered;
+import ocean.sys.SafeFork;
+import ocean.util.container.map.HashMap;
+import ocean.util.serialize.contiguous.Contiguous;
+import ocean.util.serialize.contiguous.Deserializer;
+import ocean.util.serialize.contiguous.MultiVersionDecorator;
+import ocean.util.serialize.contiguous.Serializer;
+import ocean.util.serialize.Version;
+
+/*******************************************************************************
+
+    DHT channel serializer helper.
+
+    Params:
+        S = type of record value stored in cached / mirrored channel
+
+*******************************************************************************/
+
+public class ChannelSerializer ( S )
+{
+    /// Path of the dump file.
+    private istring path;
+
+    /// Buffered output instance.
+    private BufferedOutput buffered_output;
+
+    /// Buffered input instance.
+    private BufferedInput buffered_input;
+
+    /// Fork helper for asynchronous dumping.
+    private SafeFork fork;
+
+    /// TODO
+    private alias void delegate ( void delegate ( ref hash_t, ref S ) )
+        DumpRecordDg;
```

Comment:
Might be worth supporting `Contiguous!(S)` as dump input as well, since that may well be how the data is received and stored by the application.

Reply:
Ah yes, good point. That ought to be the standard API.